tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops.td

```diff
@@ -2162,4 +2162,99 @@ def HLO_ReducePrecisionOp :
   let results = (outs HLO_FpTensor:$output);
 }
 
+def HLO_H2DOp: HLO_Op<"h2d", [NoSideEffect, SameOperandsAndResultType]> {
+  let arguments = (ins HLO_Tensor);
+  let results = (outs HLO_Tensor);
+  let hasCustomHLOConverter = 1;
+}
+
+def HLO_D2HOp: HLO_Op<"d2h", [NoSideEffect, SameOperandsAndResultType]> {
+  let arguments = (ins HLO_Tensor);
+  let results = (outs HLO_Tensor);
+  let hasCustomHLOConverter = 1;
+}
+
+def HLO_RealDynamicSliceOp: HLO_Op<
+      "dynamic_slice",
+      [NoSideEffect, AllElementTypesMatch<["operand", "result"]>,
+       AllTypesMatch<["start_indices", "limit_indices", "strides"]>]> {
```

Comment:
I'm wondering about the AllTypesMatch: that requires a strict equality between the types, but could one of them be statically sized while others have dynamic dimensions?

Reply:
Here AllTypesMatch refers to the "start_indices", "limit_indices" & "strides".
They must be a 1-D vector in the same size, which is the rank of the input.
Plus, for a dynamic rank compiler, these tensors are infact static shaped 1-D tensor.

Reply:
It may be worth adding a comment: `Expect these to be statically shaped and matching the rank of the input".

Can you add a verifier for the rank matching by the way?

Reply:
Verification func for HLO_RealDynamicSliceOp is added and update its description.  Also we add verify(HLO_DynamicPadOp) .